extras/rebalance/directory-rebalance.py

```diff
@@ -0,0 +1,203 @@
+    def generate_rebalance_file_index(self):
+        with open(self.index, 'w') as file_index:
+            total_size = 0
+            for d, _, files in os.walk(self.path):
+                for f in files:
+                    try:
+                        path = os.path.join(d, f)
+                        size = os.stat(path).st_size
+                        file_index.write(path+'-'+str(size)+'\n')
+                        self.expected_total_size += size
+                        self.expected_total_files += 1
+                        crawl_progress(self.expected_total_files, size_fmt(self.expected_total_size))
+                    except OSError as err:
+                        progress_done()
+                        print("OS error: {}".format(err))
+        progress_done()
+
+    #Stops the progress printing and prints stats collected so far
+    def exit(self):
+        progress_done()
+        rebalance_end = time.perf_counter()
+        self.duration = rebalance_end - self.rebalance_start
+
+        if self.total_files != 0:
+            print("Migrated {} / {} files [{:.2%}]".format(self.migrated_files, self.total_files, self.migrated_files/self.total_files))
+            if self.total_size != 0:
+                print("Migrated {} / {} data [{:.2%}]".format(size_fmt(self.migrated_size), size_fmt(self.total_size), self.migrated_size/self.total_size))
+        print("Run time: {}".format(time_fmt(self.duration)))
+        print("Time spent in migration: {} [{:.2%}]".format(time_fmt(self.migration_duration), self.migration_duration/self.duration))
+        print("Time spent in skipping: {} [{:.2%}]".format(time_fmt(self.skipped_migration_duration), self.skipped_migration_duration/self.duration))
+
+#/proc/mounts has active mount information. It checks that the given path is
+#mounted on glusterfs
+def check_glusterfs_supported_path(p):
+    abs_path = os.path.abspath(p)
+    if os.path.isdir(abs_path) is False:
+        raise argparse.ArgumentTypeError("{} is not a valid directory".format(abs_path))
+
+    glusterfs_mounts = {}
+    with open("/proc/mounts") as f:
+        for line in f:
+            words = line.split()
+            if len(words) < 3:
+                continue
+            if words[2] == 'fuse.glusterfs':
+                glusterfs_mounts[words[1]] = 1
+
+    p = abs_path
+    while p != '':
+        if p in glusterfs_mounts:
+            return abs_path
+        elif os.path.ismount(p):
+            raise argparse.ArgumentTypeError("{} is not a valid glusterfs path".format(abs_path))
+        else:
+            p, _, _ = p.rpartition('/')
+
+    raise argparse.ArgumentTypeError("{} is not a valid glusterfs path".format(abs_path))
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument("path", type=check_glusterfs_supported_path)
+    args = parser.parse_args()
+    r = Rebalancer(args.path)
```

Comment:
You could use `with Rebalancer(args.path) as r` and implement `__enter__` and `__exit__` methods. This way you don't need to explicitly call `exit()` function and it's guaranteed to be called in all cases (even in case of exceptions).

Reply:
Got it. This is my first real python script where I took some time to learn some concepts. Will read about this and make the required changes.